### DISCUSSION TITLE
exception in middleware

### DISCUSSION THREAD

**Discussion by BrentHuang** [fastapi/fastapi#13084] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
from collections import defaultdict
from datetime import datetime, timedelta
from fastapi import FastAPI, HTTPException, Request, status
from starlette.middleware.base import BaseHTTPMiddleware

app = FastAPI()


class RateLimitMiddleware(BaseHTTPMiddleware):

    # pylint: disable=R0913,R0917
    def __init__(self, app, max_requests: int, period: int, cleanup_interval: int = 60, client_timeout: int = 300):
        """
        :param app: FastAPI app instance
        :param max_requests: Maximum allowed requests per period
        :param period: Time window for rate limiting in seconds
        :param cleanup_interval: Interval to clean up inactive clients in seconds
        :param client_timeout: Time in seconds after which a client is considered inactive
        """
        super().__init__(app)

        self.max_requests = max_requests
        self.period = period
        self.cleanup_interval = cleanup_interval
        self.client_timeout = client_timeout
        self.rate_limit_data = defaultdict(list)  # Store client request timestamps {client_ip: [timestamps]}
        self.last_access_time = defaultdict(
            lambda: datetime.min)  # Store the last access time of each client {client_ip: last_access_time}

        asyncio.create_task(self._cleanup_task())  # Start background cleanup task

    async def dispatch(self, request: Request, call_next):
        client_ip = request.client.host
        current_time = datetime.now()

        # Update last access time
        self.last_access_time[client_ip] = current_time

        # Initialize or clean expired requests for the client
        self.rate_limit_data[client_ip] = [
            timestamp for timestamp in self.rate_limit_data[client_ip]
            if timestamp > current_time - timedelta(seconds=self.period)
        ]

        # Check if the client exceeds the rate limit
        if len(self.rate_limit_data[client_ip]) >= self.max_requests:
            raise HTTPException(status_code=status.HTTP_429_TOO_MANY_REQUESTS, detail="Too Many Requests")

        # Record the current request
        self.rate_limit_data[client_ip].append(current_time)

        # Proceed with the request
        response = await call_next(request)
        return response

    async def _cleanup_task(self):
        """Background task to clean up inactive clients."""
        while True:
            await asyncio.sleep(self.cleanup_interval)
            current_time = datetime.now()

            # Identify clients to remove based on inactivity
            inactive_clients = [
                client_ip for client_ip, last_access in self.last_access_time.items()
                if current_time - last_access > timedelta(seconds=self.client_timeout)
            ]
            # Remove inactive clients
            for client_ip in inactive_clients:
                del self.rate_limit_data[client_ip]
                del self.last_access_time[client_ip]


app.add_middleware(RateLimitMiddleware, max_requests=5, period=60, cleanup_interval=30, client_timeout=300)


@app.get('/')
async def root():
    return {'message': 'hello, world!'}


# uvicorn x:app --host 0.0.0.0 --port 8000 --reload
```

### Description

      INFO   Application startup complete.
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 404
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 404
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 404
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 404
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 404
      INFO   127.0.0.1:55328 - "GET /unauthorized HTTP/1.1" 500
     ERROR   Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/_utils.py", line 76, in collapse_excgroups
  |     yield
  |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 186, in __call__
  |     async with anyio.create_task_group() as task_group:
  |                ^^^^^^^^^^^^^^^^^^^^^^^^^
  |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 815, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
    |     result = await app(  # type: ignore[func-returns-value]
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    |     return await self.app(scope, receive, send)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    |     raise exc
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 185, in __call__
    |     with collapse_excgroups():
    |          ^^^^^^^^^^^^^^^^^^^^
    |   File "/home/hgc/.pyenv/versions/3.12.8/lib/python3.12/contextlib.py", line 158, in __exit__
    |     self.gen.throw(value)
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    |     raise exc
    |   File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 187, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/hgc/cxy/farm_server/x.py", line 48, in dispatch
    |     raise HTTPException(status_code=status.HTTP_429_TOO_MANY_REQUESTS, detail="Too Many Requests")
    | fastapi.exceptions.HTTPException: 429: Too Many Requests
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 185, in __call__
    with collapse_excgroups():
         ^^^^^^^^^^^^^^^^^^^^
  File "/home/hgc/.pyenv/versions/3.12.8/lib/python3.12/contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    raise exc
  File "/home/hgc/cxy/farm_server/.venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 187, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/hgc/cxy/farm_server/x.py", line 48, in dispatch
    raise HTTPException(status_code=status.HTTP_429_TOO_MANY_REQUESTS, detail="Too Many Requests")
fastapi.exceptions.HTTPException: 429: Too Many Requests

### Operating System

Linux

### FastAPI Version

0.115.6

### Pydantic Version

2.10.3

### Python Version

Python 3.12.8

**Comment by yvallois** [comment]
Your code seems to return a 500, you should return something like `JSONResponse({“detail”: “Too Many Requests”}, status_code=status.HTTP_429_TOO_MANY_REQUESTS)` rather than raising an error if you want to see the 429

**Reply by yvallois** [reply]
you can also use [pure-asgi-middleware](https://www.starlette.io/middleware/#pure-asgi-middleware). In this case, you can look at the code at https://github.com/abersheeran/asgi-ratelimit for inspiration.

**Comment by PREPONDERANCE** [comment]
I believe you should add a exception handler to fastapi so that it can capture this `429 Too many requests` error and handle it the way you want.

```Python
from fastapi import HTTPException, Request

async def custom_handler_function(request: Request, exception: HTTPException):
    # Handle 429 or other errors, you can obtain this with `exception.status_code`
    pass

app.add_exception_handler(HTTPException, custom_handler_function)
```

**Reply by nbx3** [reply]
Even with a custom handler the unhandled errors in a TaskGroup still occurs

**Comment by nbx3** [comment]
It seems exceptions raised in middleware cannot be caught, here's an easier to test example

```python
from typing import Callable

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.exceptions import HTTPException as StarletteHTTPException

from fastapi import FastAPI, Request, HTTPException
from fastapi.responses import PlainTextResponse

class TestMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next: Callable):
        if request.url.path.startswith('/1'):
            raise HTTPException(status_code=403, detail="This is a 403 unauthorised exception")
        return await call_next(request)


app = FastAPI()
app.add_middleware(TestMiddleware)

@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request: Request, exc: StarletteHTTPException):
    return PlainTextResponse(f"Detail: {exc.detail}", status_code=exc.status_code)

@app.get("/1")
async def ex1():
    return PlainTextResponse(f"Example will produce uncaught exception")

@app.get("/2")
async def ex2():
    HTTPException(status_code=400, detail="Example will produce a caught exception")
    return PlainTextResponse(f"Example will produce a caught exception")
```

Calling `/2` raises an exception and is caught as you would expect by the exception handler. Calling `\1` produces 

```
  + Exception Group Traceback (most recent call last):
  |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/_utils.py", line 76, in collapse_excgroups
  |     yield
  |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 186, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 815, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
    |     result = await app(  # type: ignore[func-returns-value]
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    |     return await self.app(scope, receive, send)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    |     raise exc
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 185, in __call__
    |     with collapse_excgroups():
    |   File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    |     self.gen.throw(value)
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    |     raise exc
    |   File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 187, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/home/nick/Projects/fastapi/example.py", line 12, in dispatch
    |     raise HTTPException(status_code=403, detail="This is a 403 unauthorised exception")
    | fastapi.exceptions.HTTPException: 403: This is a 403 unauthorised exception
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 409, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 185, in __call__
    with collapse_excgroups():
  File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    raise exc
  File "/home/nick/Projects/fastapi/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 187, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/nick/Projects/fastapi/example.py", line 12, in dispatch
    raise HTTPException(status_code=403, detail="This is a 403 unauthorised exception")
fastapi.exceptions.HTTPException: 403: This is a 403 unauthorised exception
```

**Reply by Kludex** [reply, marked as answer]
Stop sending the same message on every discussion, please.

That's correct. It's as intended.

**Reply by nbx3** [reply]
It's intended to break in this scenario? Can you explain how to catch an exception raised in middleware?

**Reply by Kludex** [reply]
The exception handlers only handle exceptions from the endpoints, not on middlewares.

You can send a JSONResponse from the middleware instead.

**Reply by nbx3** [reply]
Thanks I'll try that though it feels a bit of a workaround.

**Reply by Kludex** [reply]
You can read more about the order of middleware here: https://www.starlette.io/middleware/

But the point is that your middleware runs on an outer scope considering the ExceptionMiddleare (that handles exceptions in your endpoints).

It's not a workaround. It's what you should do.

**Reply by nbx3** [reply]
So its possible to return a PlaintextResponse or JSONResponse but not a TemplateReponse, I get the same type of "unhandled errors in taskgroup". Is this a known issue?

**Comment by lionpeloux** [comment]
Just adding this MWE as I struggled few hours too on this topic, might help others.
Maybe a small paragraph in the Testing doc to make that explicit could help ?

```python
from fastapi import FastAPI, HTTPException, Request
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware

app = FastAPI()

class CustomMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        # Example condition to raise an HTTP exception
        if "Authorization" not in request.headers:
            return JSONResponse(status_code=401, content={"detail": "Not authenticated"})
            # raise HTTPException(status_code=401, detail="Not authenticated")

        response = await call_next(request)
        return response
    
app.add_middleware(CustomMiddleware)


@app.get("/")
async def read_root():
    raise HTTPException(status_code=401, detail="Raised exception")

# TESTING

from fastapi.testclient import TestClient
client = TestClient(app)

def test_raise_with_header():
    response = client.get("/", headers={"Authorization": "Bearer token"})
    assert response.status_code == 401
    assert response.json() == {"detail": "Raised exception"}

def test_raise_no_header_with_jsonresponse():
    response = client.get("/")
    assert response.status_code == 401
    assert response.json() == {"detail": "Not authenticated"}

# def test_raise_no_header_with_httpexception():
#     try:
#         response = client.get("/")
#     except HTTPException as e:
#         assert e.status_code == 401
#         assert e.detail == "Not authenticated"
```
